File: Data Structures & Algorithms/two-integer-sum-ii/submission-7.java
class Solution {
    public int[] twoSum(int[] numbers, int target) {
        int l = 0, r = numbers.length - 1;
        //O( n log(n) + log(n)) -> n log(n)
        
        while(l < r) {
            //int mid = l + (r - l) / 2;
            int sum = numbers[l] + numbers[r];
            if(sum == target) return new int[]{l+1, r+1};
            else if(sum < target) l++;
            else r--; 
        }

        return new int[]{-1, -1};
    }
}

File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-7.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        res = -1
        l, r = 0, len(nums) - 1
        while l <= r:
            mid = l + (r - l) // 2
            if nums[mid] == target:
                return mid

            # the left half is sorted
            if nums[l] <= nums[mid]:
                # value is below the mid point
                if nums[l] <= target and target < nums[mid]:
                    r = mid - 1
                # else value is above the mid point (either nums[l] > target or nums[mid] < target)
                else:
                    l = mid + 1

            # the right half is sorted (nums[l] > nums[mid])
            else:
                # move left pointer
                if nums[l] > target and nums[mid] <= target:
                    l = mid + 1
                else:
                    r = mid - 1
        return res